src/timestamp.rs

```diff
@@ -6,26 +6,50 @@
 
 use hlc;
 use std::sync::Mutex;
+use memmap::{MmapViewSync, Mmap, Protection};
 
-lazy_static! {
-
-   static ref HLC_CLOCK: Mutex<hlc::Clock<hlc::Wall>> = Mutex::new(hlc::Clock::wall());
-
-}
-
-pub fn hlc() -> hlc::Timestamp<hlc::WallT> {
-    (*HLC_CLOCK).lock().unwrap().now()
+#[derive(Debug)]
+pub struct Timestamp {
+    clock: Mutex<(hlc::Clock<hlc::Wall>, MmapViewSync)>,
 }
 
-#[cfg(test)]
-mod tests {
-
-    use timestamp;
+impl Timestamp {
+    /// Create a new Timestamp clock. First the passed in memory map will be checked to check if
+    /// a previous timestamp exists. If one exists (i.e. if the results aren't 20 bytes of 0) it
+    /// will be "observed" by the HLC library.
+    pub fn new(scratchpad: Option<MmapViewSync>) -> Self {
+        if scratchpad.is_some() {
+            let scratchpad = scratchpad.unwrap();
+            let clock = {
+                let mut clock = hlc::Clock::wall();
+                let previous = unsafe { scratchpad.as_slice() };
+                let res = hlc::Timestamp::read_bytes(previous);
+                if res.is_ok() {
+                    let res = res.unwrap();
+                    if res > clock.now() {
+                        let _ = clock.observe(&res);
+                    }
+                }
+                clock
+            };
+            Timestamp {
+                clock: Mutex::new((clock, scratchpad)),
+            }
+        } else {
+            let clock = hlc::Clock::wall();
+            let scratchpad = Mmap::anonymous(20, Protection::ReadWrite).unwrap();
+            Timestamp {
+                clock: Mutex::new((clock, scratchpad.into_view_sync()))
+            }
+        }
+    }
 
-    #[test]
-    fn test() {
-        let t1 = timestamp::hlc();
-        let t2 = timestamp::hlc();
-        assert!(t2 > t1);
+    pub fn hlc(&self) -> hlc::Timestamp<hlc::WallT> {
+        let mut clock = self.clock.lock().unwrap();
+        let now = clock.0.now();
+        let ref mut state = clock.1;
+        let _ = unsafe { &now.write_bytes(&mut state.as_mut_slice()).unwrap() };
```

Comment:
Right now there is no guarantee when it's flushed. This is unacceptable for a production system. We should look into using [`flush_async`](https://osnacs.github.io/plantex/memmap/struct.MmapViewSync.html#method.flush_async) but I'd like that to be done later, preferably with a benchmark on the HLC.